src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java

```diff
@@ -0,0 +1,472 @@
+        }
+
+        if (returningSelect != null)
+        {
+            TxnNamedRead read = createNamedRead(returningSelect, options);
+            keyConsumer.accept(read.key());
+            reads.add(read);
+        }
+
+        for (NamedSelect select : autoReads.values())
+            // don't need keyConsumer as the keys are known to exist due to Modification
+            reads.add(createNamedRead(select, options));
+        
+        return new TxnRead(reads);
+    }
+
+    TxnCondition createCondition(QueryOptions options)
+    {
+        if (conditions.isEmpty())
+            return TxnCondition.none();
+        if (conditions.size() == 1)
+            return conditions.get(0).createCondition(options);
+
+        List<TxnCondition> result = new ArrayList<>(conditions.size());
+        for (ConditionStatement condition : conditions)
+            result.add(condition.createCondition(options));
+
+        // TODO: OR support
+        return new TxnCondition.BooleanGroup(TxnCondition.Kind.AND, result);
+    }
+
+    private final Map<TxnDataName, NamedSelect> autoReads = new HashMap<>();
+
+    List<TxnWrite.Fragment> createWriteFragments(ClientState state, QueryOptions options, Consumer<Key> keyConsumer)
+    {
+        List<TxnWrite.Fragment> fragments = new ArrayList<>(updates.size());
+        int idx = 0;
+        for (ModificationStatement modification : updates)
+        {
+            TxnWrite.Fragment fragment = modification.getTxnWriteFragment(idx++, state, options);
+            keyConsumer.accept(fragment.key);
+            fragments.add(fragment);
+
+            if (modification.allReferenceOperations().stream().anyMatch(ReferenceOperation::requiresRead))
+            {
+                TxnDataName partitionName = TxnDataName.partitionRead(modification.metadata(), fragment.key.partitionKey());
+                if (!autoReads.containsKey(partitionName))
+                    autoReads.put(partitionName, new NamedSelect(partitionName, ((UpdateStatement) modification).createSelectForTxn()));
+            }
+        }
+        return fragments;
+    }
+
+    TxnUpdate createUpdate(ClientState state, QueryOptions options, Consumer<Key> keyConsumer)
+    {
+        return new TxnUpdate(createWriteFragments(state, options, keyConsumer), createCondition(options));
+    }
+
+    Keys toKeys(SortedSet<Key> keySet)
+    {
+        return new Keys(keySet);
+    }
+
+    @VisibleForTesting
+    public Txn createTxn(ClientState state, QueryOptions options)
+    {
+        SortedSet<Key> keySet = new TreeSet<>();
+        if (updates.isEmpty())
+        {
+            Preconditions.checkState(conditions.isEmpty());
+            TxnRead read = createRead(options, keySet::add);
+            return new Txn.InMemory(toKeys(keySet), read, TxnQuery.ALL);
+        }
+        else
+        {
+            TxnUpdate update = createUpdate(state, options, keySet::add);
+            TxnRead read = createRead(options, keySet::add);
```

Comment:
Could `createRead` be pulled out of the two branches since it is done the same in both?